jupyterhub_chameleon/trovi.py

```diff
@@ -0,0 +1,88 @@
+import logging
+import os
+from urllib.parse import urljoin
+
+import requests
+from tornado.web import HTTPError
+
+LOG = logging.getLogger(__name__)
+
+def exchange_token(access_token):
+    # Exchange the user's keystone token for a Trovi token
+    token_resp = requests.post(
+        urljoin(os.getenv("TROVI_URL"), "/token/"),
+        headers={
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        },
+        json={
+            "grant_type": "token_exchange",
+            "subject_token": access_token,
+            "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
+            "scope": "artifacts:read artifacts:write",
+        },
+    )
+    LOG.debug(f"Trovi token exchange response: {token_resp}")
+
+    if token_resp.status_code in (
+            requests.codes.unauthorized,
+            requests.codes.forbidden,
+    ):
+        LOG.error(f"Authentication to trovi failed: {token_resp}")
+        raise HTTPError(
+            requests.codes.unauthorized,
+            "You are not authorized to upload artifacts to Trovi via Jupyter.",
+        )
+    elif token_resp.status_code != requests.codes.created:
+        LOG.error(f"Authentication to trovi failed: {token_resp}")
+        raise HTTPError(
+            requests.codes.internal_server_error,
+            requests.codes.internal_server_error,
+            "Unknown error uploading artifact to Trovi.",
+        )
+
+    return token_resp.json()
+
+def fetch_artifact(artifact_id, trovi_token):
+    artifact_req = requests.PreparedRequest()
+    artifact_req.prepare_url(
+        urljoin(os.getenv("TROVI_URL"), f"/artifacts/{artifact_id}"),
+        {"access_token": trovi_token["access_token"]},
```

Comment:
I thought we were using `Authentication` header to pass this to Trovi, but I looked back at the design doc and realized I didn't specify this. As such, this is ok, but not standard practice for security reasons.

See https://datatracker.ietf.org/doc/html/rfc6750#section-5.3 for why. We have a bit more control over where the URL is handled so it's not a big deal, but good for future knowledge.